httprequest_lego_provider/tests/test_dns.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""Unit tests for the dns module."""
+
+import secrets
+from pathlib import Path
+from unittest.mock import MagicMock, patch
+
+import pytest
+from git import GitCommandError, Repo
+
+from httprequest_lego_provider.dns import DnsSourceUpdateError, remove_dns_record, write_dns_record
+
+
+@patch.object(Path, "write_text")
+@patch.object(Repo, "clone_from")
+def test_write_dns_record_raises_exception(repo_patch, _):
+    """
+    arrange: mock the repo so that it raises a GitCommandError.
+    act: attempt to write a new DNS record.
+    assert: a DnsSourceUpdateError exception is raised.
+    """
+    repo_patch.side_effect = GitCommandError("Error executing command")
+
+    dns_record = "site.example.com"
+    with pytest.raises(DnsSourceUpdateError):
+        write_dns_record(dns_record, secrets.token_hex())
+
+
+@patch.object(Path, "write_text")
+@patch.object(Path, "read_text")
+@patch.object(Repo, "clone_from")
+def test_write_dns_record(repo_patch, read_patch, write_patch):
+    """
+    arrange: mock the repo.
+    act: attempt to write a new DNS record.
+    assert: a new file with filename matching the record is committed and pushed to the repository.
+    """
+    repo_mock = MagicMock(spec=Repo)
+    repo_patch.return_value = repo_mock
+    token = secrets.token_hex()
+    read_patch.return_value = (
+        "site 600 IN TXT \042sometoken\042\n"
+        "site1 600 IN TXT \042sometoken\042\n"
+        "site3 600 IN TXT \042sometoken\042\n"
+    )
+
+    dns_record = "site.example.com"
```

Comment:
I would parametrize the test and pass different scenarios:
- simple tld
- domain.tld
- subdomain.domain.tld
- long.long.long.long.subdomain.chain.domain.tld
- Subdomain and domains with IDN (see https://en.wikipedia.org/wiki/Internationalized_domain_name)

If we implement the check for "_acme_challenge" I'd also implement a test with and without "_acme_challenge"